### DISCUSSION TITLE
Can I make microblog posts without titles?

### DISCUSSION THREAD

**Discussion by eli-oat** [datenstrom/community#299] [Ask a question] [answered]
Currently, when a user creates a blog post without a title the title is displayed as "None." 

Rather than this, I believe the title should be completely omitted (wordpress uses the post-date as the title when none is provided). 

I'm interested in writing a [micropub endpoint](https://indieweb.org/micropub-endpoint) plugin for Yellow to make it indieweb compatible, as well as compatible with Micro.Blog. Support for posts without titles would help this effort.

**Comment by BsNoSi** [comment]
I have to admit, that for me the current method is one of the major advantages of Yellow in compare to others. It is extremly simple to create and submit a new entry with a markdown file. 

As far as I understand *micropub*, this targets a completely different needs. In compare to the current method of editing directly in the webpage or upload a file via FTP or directly with an editor, it is a quite *heavyweight* procedure.

I can not see an advantage, to have Yellow "indieweb compatible“ – Yellow is "Yellow compatible", driven by the same ideas as indieweb:  [owning your domain, using it as your primary identity to publish on your own site (optionally syndicate elsewhere), and owning your data](https://indieweb.org/IndieWeb "taken from indiweb homepage")".

I am sorry, but for me there is no need for this plugin, bcause the explained targets are already been reached. But my opinion is not relevant. If you like to create a plugin for that, you should do it. The only thing that you should resepect are the ideas behind Yellow, instead of expecting Yellow has to change for your needs.

I believe, if you think a little bit about the challenge, you will find a solution for „untitled pages“ – for me it´s obvious.

**Comment by markseuffert** [comment, marked as answer]
Yes, you can use the [blog extension](https://github.com/annaesvensson/yellow-blog) with a timestamp as a substitute title.

Here's an example file `content/shared/page-new-blog.md`:

```
---
Title: @timestamp
Published: @datetime
Author: @username
Layout: blog
Tag: microblog
---
Write text here...
```

Create a new blog page in the web browser. Log in with your user account, click `+` at the top right, then click `Create` and you should automatically get the correct page settings. The title is used to generate the file name and the URL of a page. The HTML code of a page can be customised in the layout file. Have a look at the existing layout files and customise them.  For example remove the line `<div class="entry-title">...</div>` from file `system/layouts/blog-start.html`.

Tested with Datenstrom Yellow 0.8.23. Hope this helps.

**Comment by eli-oat** [comment]
@markseu I've been exploring what it would take to build a [micropub](https://micropub.net) endpoint for Yellow. This would allow folks to post to Yellow sites directly through the existing interface, as well as through a variety of micropub clients, like [Quill](https://quill.p3k.io) and [Micro.Blog](https://micro.blog)'s iOS app. 

The primary reason to do this is to use your personal site as a micro blog, micro blog posts (like tweets) feature a short snippet of text, and no post title. 

Making the micropub endpoint is relatively straight forward (I actually already have a working proof of concept in place), what I'm looking for now is a way to make blog post titles optional — right now Yellow seems to fill the title with "NONE" or a number if no title is provided. I'm wondering if there is a better way to handle blank titles? 

**[EDITED]**: Realized I had an invisible character in the template that was throwing the error. Your `@timestamp` suggestion seems to be working!

**Reply by pftnhr** [reply]
Based on Mark's code in [#372](https://github.com/datenstrom/yellow/discussions/372#discussioncomment-244418)  I've created an extension that sets the first 25 characters of the content, followed by the date of publication in the format `Ymd` as `titleSlug`.

``` php
<?php
// Transliterate extension, experimental

class YellowTransliterate {
    const VERSION = "0.8.17";
    public $yellow;         //access to API

    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
    }
    
    // Handle content file editing
    public function onEditContentFile($page, $action, $email) {
        if (!$page->isExisting("titleSlug")) {
            $replaceData = $this->getReplaceData();
            $titleSlug = substru(str_replace(array_keys($replaceData), array_values($replaceData), $page->getContentRaw()), 0, 25).$page->getDate("published", "Ymd");
            $titleSlug = strtoloweru(preg_replace("/[^\pL\d\-]+/u", "-", $titleSlug));
            $titleSlug = preg_replace("/^-+|-+$/", "", $titleSlug);
            $page->rawData = $this->yellow->toolbox->setMetaData($page->rawData, "titleSlug", $titleSlug);
        }
    }
    
    // Return text replace data, UTF8 to ASCII
    public function getReplaceData() {
        return array(
            "À" => "A", "Á" => "A", "Â" => "A", "Ã" => "A", "Ä" => "Ae", "Å" => "A", "Æ" => "AE", "Ç" => "C", 
            "È" => "E", "É" => "E", "Ê" => "E", "Ë" => "E", "Ì" => "I", "Í" => "I", "Î" => "I", "Ï" => "I", 
            "Ð" => "D", "Ñ" => "N", "Ò" => "O", "Ó" => "O", "Ô" => "O", "Õ" => "O", "Ö" => "Oe", "Ő" => "O", 
            "Ø" => "O", "Ù" => "U", "Ú" => "U", "Û" => "U", "Ü" => "Ue", "Ű" => "U", "Ý" => "Y", "Þ" => "TH", 
            "ß" => "ss", 
            "à" => "a", "á" => "a", "â" => "a", "ã" => "a", "ä" => "ae", "å" => "a", "æ" => "ae", "ç" => "c", 
            "è" => "e", "é" => "e", "ê" => "e", "ë" => "e", "ì" => "i", "í" => "i", "î" => "i", "ï" => "i", 
            "ð" => "d", "ñ" => "n", "ò" => "o", "ó" => "o", "ô" => "o", "õ" => "o", "ö" => "oe", "ő" => "o", 
            "ø" => "o", "ù" => "u", "ú" => "u", "û" => "u", "ü" => "ue", "ű" => "u", "ý" => "y", "þ" => "th", 
            "ÿ" => "y",
            // Add more characters here, see also: 
            // https://gist.github.com/sgmurphy/3098978
            // https://github.com/ausi/slug-generator/blob/master/src/Resources/Latin-ASCII.txt
            // https://github.com/jbroadway/urlify/blob/master/URLify.php
        );
    }
}
```

The `@timestamp` is used as `title`. This allows you to remove the title from the layout files and it is no longer needed anywhere else.

**Comment by BsNoSi** [comment]
I had a short look on "Quill" and "Micro Blog" and have no reasonable idea, why I should hand out my login data (that shurely are requried) to a third party instead to log in directly.

But the idea, to have a simple interface for fast blogging is somehow interesting. Would it not be better/easier, to have an app on your computer/mobile that allows you to post without a web interface?

Such an app may have a „create header entries“ button for creating/sending valid entries that are required.

I will/can not promise anything, but if there is an interest in such an app, I may think about it.

**Comment by schulle4u** [comment]
Another microblogging solution which could be adapted to work with Yellow is [TWTXT](https://github.com/buckket/twtxt), however this is rather geeky and requires shell access to post updates. An advantage of this solution is that you don't need to access a serverside API, the posts are stored in a simple text file which can be pushed to the server using FTP or SFTP. Using [a CSV parser for php](https://github.com/parsecsv/parsecsv-for-php), you can parse this file and turn it into a nice looking microblog site for Yellow. 

A general CSV parser plugin already is on my to-do-list, I can try to create this if anyone is interested. 

Steffen

**Comment by eli-oat** [comment]
@schulle4u I like the sound of that a lot! 

@BnNoSi I certainly understand hesitancy to hand out login credentials, but that isn’t required if a site has a micropub endpoint enabled. 

Micropub clients rely on a system called indieauth to handle user authentication. There is no centeralized login server this way, so there is no need to share login credentials! 

Here is some more info on indieauth: https://indieweb.org/IndieAuth

**Comment by BsNoSi** [comment]
Does this mean that it is enough to get access to this single account to get access to all accounts served by this mechanism?

Since such a service is of great interest for hackers because fo the booty is many times bigger, the risk of having uninvited guests on my domain increases significantly.
 
I have to admit, that I still can not see an advantage on that. Is it really that complicated to log into Yellow for a post?

**Comment by eli-oat** [comment]
@BsNoSi I totally understand your concern, but indieauth isn't a single point of entry. Rather, it is a standard that allows folks to use their own domain name as credentials to sign in across the web. 

With this setup, there is no primary attack vector for hackers because the attack surface would be the specific domain name itself. There is no password to brute force, or guess. Instead, tokens are generated against the domain name. For more info, check out [indieauth.com](https://indieauth.com)

While I understand that you don't see a value in this, it is something I'm interested in exploring — and not asking to be baked into Yellow's core, by any means. Instead, I'm looking to create a plug in to support these features. There is an vibrant community of folks involved in the Indieweb, and think Yellow could be a valuable tool for them. 

The primary value that I see in supporting micropub (and indieauth) is allowing users to [create posts within different context, such as replies and likes](https://indieweb.org/posts#Kinds_of_Posts).
